cardano-cli/src/Cardano/CLI/Ping/Lib.hs

```diff
@@ -0,0 +1,463 @@
+toStatPoint ts host cookie sample td =
+  StatPoint
+    { spTimestamp = ts
+    , spHost      = host
+    , spCookie    = cookie
+    , spSample    = sample
+    , spMedian    = quantile' 0.5
+    , spP90       = quantile' 0.9
+    , spMean      = mean'
+    , spMin       = minimumValue td
+    , spMax       = maximumValue td
+    , spStd       = stddev'
+    }
+  where
+    quantile' :: Double -> Double
+    quantile' q = fromMaybe 0 (quantile q td)
+
+    mean' :: Double
+    mean' = fromMaybe 0 (mean td)
+
+    stddev' :: Double
+    stddev' = fromMaybe 0 (stddev td)
+
+pingClient :: Tracer IO LogMsg -> PingCmd -> [NodeVersion] -> AddrInfo -> IO ()
+pingClient tracer PingCmd{pingCmdQuiet, pingCmdJson, pingCmdCount} versions peer = bracket
+  (Socket.socket (Socket.addrFamily peer) Socket.Stream Socket.defaultProtocol)
+  Socket.close
+  (\sd -> withTimeoutSerial $ \timeoutfn -> do
+    when (Socket.addrFamily peer /= Socket.AF_UNIX) $ do
+      Socket.setSocketOption sd Socket.NoDelay 1
+      Socket.setSockOpt sd Socket.Linger $ StructLinger
+        { sl_onoff  = 1
+        , sl_linger = 0
+        }
+
+    !t0_s <- getMonotonicTime
+    Socket.connect sd (Socket.addrAddress peer)
+    !t0_e <- getMonotonicTime
+    peerStr <- peerString
+    unless pingCmdQuiet $ printf "%s network rtt: %.3f\n" peerStr $ toSample t0_e t0_s
+
+    let timeout = 30
+        bearer = socketAsMuxBearer timeout nullTracer sd
+
+    !t1_s <- write bearer timeoutfn $ wrap handshakeNum InitiatorDir (handshakeReq versions)
+    (msg, !t1_e) <- nextMsg bearer timeoutfn handshakeNum
+    unless pingCmdQuiet $ printf "%s handshake rtt: %s\n" peerStr (show $ diffTime t1_e t1_s)
+
+    case CBOR.deserialiseFromBytes handshakeDec msg of
+      Left err -> eprint $ printf "%s Decoding error %s\n" peerStr (show err)
+      Right (_, Left err) -> eprint $ printf "%s Protocol error %s\n" peerStr (show err)
+      Right (_, Right version) -> do
+        unless pingCmdQuiet $ printf "%s Negotiated version %s\n" peerStr (show version)
+        keepAlive bearer timeoutfn peerStr version (tdigest []) 0
+        -- send terminating message
+        _ <- write bearer timeoutfn $ wrap keepaliveNum InitiatorDir (keepAliveDone version)
+        -- protocol idle timeout
+        MT.threadDelay 5
+
+  )
+  where
+
+    peerString :: IO String
+    peerString =
+      case Socket.addrFamily peer of
+        Socket.AF_UNIX -> return $ show $ Socket.addrAddress peer
+        _ -> do
+          (Just host, Just port) <-
+            Socket.getNameInfo
+              [Socket.NI_NUMERICHOST, Socket.NI_NUMERICSERV]
+              True True (Socket.addrAddress peer)
+          return $ host <> ":" <> port
+
+    toSample :: Time -> Time -> Double
+    toSample t_e t_s = realToFrac $ diffTime t_e t_s
+
+    eprint :: String -> IO ()
+    eprint = BS.Char.hPutStr IO.stderr . BS.Char.pack
+
+    nextMsg ::  MuxBearer IO -> TimeoutFn IO -> MiniProtocolNum -> IO (LBS.ByteString, Time)
```

Comment:
A comment describing this function would be useful

Reply:
The code is now in `ouroboros-network`.